src/ess/diffraction/types.py

```diff
@@ -0,0 +1,118 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+
+"""This module defines the domain types uses in ess.diffraction.
+
+The domain types are used to define parameters and to request results from a Sciline
+pipeline.
+"""
+
+from typing import NewType, TypeVar
+
+import sciline
+import scipp as sc
+
+# 1 TypeVars used to parametrize the generic parts of the workflow
+
+# 1.1 Run types
+EmptyInstrumentRun = NewType('EmptyInstrumentRun', int)
+"""Empty instrument run."""
+SampleRun = NewType('SampleRun', int)
+"""Sample run."""
+VanadiumRun = NewType('VanadiumRun', int)
+"""Vanadium run."""
+RunType = TypeVar('RunType', EmptyInstrumentRun, SampleRun, VanadiumRun)
+"""TypeVar used for specifying the run."""
+
+
+# 2 Workflow parameters
+
+CalibrationFilename = NewType('CalibrationFilename', str)
+"""Filename of the instrument calibration file."""
+
+DspacingBins = NewType('DSpacingBins', sc.Variable)
+"""Bin edges for d-spacing."""
+
+
+class Filename(sciline.Scope[RunType, str], str):
+    """Filename of a run."""
+
+
+OutFilename = NewType('OutFilename', str)
+"""Filename of the output."""
+
+TwoThetaBins = NewType('TwoThetaBins', sc.Variable)
+"""Bin edges for 2theta."""
```

Comment:
I suppose this is for grouping by scattering angle, instead of into a single spectrum? Can you document this?